compiler/rustc_trait_selection/src/errors.rs

```diff
@@ -31,23 +32,50 @@ pub struct UnableToConstructConstantValue<'a> {
 }
 
 #[derive(Diagnostic)]
-#[diag(trait_selection_empty_on_clause_in_rustc_on_unimplemented, code = E0232)]
-pub struct EmptyOnClauseInOnUnimplemented {
-    #[primary_span]
-    #[label]
-    pub span: Span,
-}
-
-#[derive(Diagnostic)]
-#[diag(trait_selection_invalid_on_clause_in_rustc_on_unimplemented, code = E0232)]
-pub struct InvalidOnClauseInOnUnimplemented {
-    #[primary_span]
-    #[label]
-    pub span: Span,
+pub enum InvalidOnClause {
+    #[diag(trait_selection_rustc_on_unimplemented_empty_on_clause, code = E0232)]
+    Empty {
+        #[primary_span]
+        #[label]
+        span: Span,
+    },
+    #[diag(trait_selection_rustc_on_unimplemented_expected_one_predicate_in_not, code = E0232)]
+    ExpectedOnePredInNot {
+        #[primary_span]
+        #[label]
+        span: Span,
+    },
+    #[diag(trait_selection_rustc_on_unimplemented_on_clause_literal, code = E0232)]
+    UnSupportedLiteral {
```

Comment:
s/UnSupported/Unsupported/

Reply:
Also, the variant name doesn't really match the message name: "unsupported" vs "unimplemented". It would be better if they were close to each other. Likewise for a couple of the other variants.

Reply:
I'm not sure what you mean. The `unimplemented` in the flt name references the name of the attribute, not what went wrong. But I have edited it to `...unsupported_literal_in_on`

Reply:
Maybe consider having `trait_selection_rou_*` rather than `trait_selection_rustc_on_unimplemented_*`? I errored on the side of not abbreviating.

Reply:
Generally the name of the error type and the name of the error message are similar, as is the text of the error. Here are some case unlike that. For example:
```
trait_selection_rustc_on_unimplemented_invalid_boolean
"invalid boolean flag"
InvalidFlag
```
One uses "boolean", one uses "flag", one uses "boolean flag". More consistency would help.

Reply:
I hadn't really thought about that, thank you. I've made some changes to the naming.